modules/light-clients/08-wasm/types/querier.go

```diff
@@ -0,0 +1,150 @@
+	sdkGas := VMGasRegister.FromWasmVMGas(gasLimit)
+
+	subCtx, _ := q.Ctx.WithGasMeter(storetypes.NewGasMeter(sdkGas)).CacheContext()
+
+	// make sure we charge the higher level context even on panic
+	defer func() {
+		q.Ctx.GasMeter().ConsumeGas(subCtx.GasMeter().GasConsumed(), "contract sub-query")
+	}()
+
+	if request.Stargate != nil {
+		return GetQueryPlugins().Stargate(subCtx, request.Stargate)
+	}
+
+	if request.Custom != nil {
+		return GetQueryPlugins().Custom(subCtx, request.Custom)
+	}
+
+	return nil, wasmvmtypes.UnsupportedRequest{Kind: "Unsupported query request"}
+}
+
+type (
+	CustomQuerier   func(ctx sdk.Context, request json.RawMessage) ([]byte, error)
+	StargateQuerier func(ctx sdk.Context, request *wasmvmtypes.StargateQuery) ([]byte, error)
+)
+
+// QueryPlugins is a list of query handlers that can be used to extend the default querier.
+type QueryPlugins struct {
+	Custom   CustomQuerier
+	Stargate StargateQuerier
+}
+
+// Merge merges the query plugin with a provided one.
+func (e QueryPlugins) Merge(x *QueryPlugins) QueryPlugins {
+	// only update if this is non-nil and then only set values
+	if x == nil {
+		return e
+	}
+
+	if x.Custom != nil {
+		e.Custom = x.Custom
+	}
+
+	if x.Stargate != nil {
+		e.Stargate = x.Stargate
+	}
+
+	return e
+}
+
+// SetQueryPlugins sets the current query plugins
+func SetQueryPlugins(plugins *QueryPlugins) {
+	queryPlugins = plugins
+}
+
+// GetQueryPlugins returns the current query plugins
+func GetQueryPlugins() *QueryPlugins {
+	return queryPlugins
+}
+
+// NewDefaultQueryPlugins returns the default set of query plugins
+func NewDefaultQueryPlugins() *QueryPlugins {
+	return &QueryPlugins{
+		Custom:   RejectCustomQuerier(),
+		Stargate: AcceptListStargateQuerier([]string{}),
+	}
+}
+
+// AcceptListStargateQuerier allows all queries that are in the accept list provided and in the default accept list.
+// This function returns protobuf encoded responses.
+func AcceptListStargateQuerier(accepted []string) func(sdk.Context, *wasmvmtypes.StargateQuery) ([]byte, error) {
+	return func(ctx sdk.Context, request *wasmvmtypes.StargateQuery) ([]byte, error) {
+		// A default list of accepted queries can be added here.
+		// accepted = append(defaultAcceptList, accepted...)
+
+		isAccepted := slices.Contains(accepted, request.Path)
+		if !isAccepted {
+			return nil, wasmvmtypes.UnsupportedRequest{Kind: fmt.Sprintf("'%s' path is not allowed from the contract", request.Path)}
+		}
+
+		route := ibcwasm.GetQueryRouter().Route(request.Path)
+		if route == nil {
+			return nil, wasmvmtypes.UnsupportedRequest{Kind: fmt.Sprintf("No route to query '%s'", request.Path)}
+		}
+
+		res, err := route(ctx, &abci.RequestQuery{
+			Data: request.Data,
+			Path: request.Path,
+		})
+		if err != nil {
+			return nil, err
+		}
+		if res == nil || res.Value == nil {
+			return nil, errorsmod.Wrap(ErrInvalid, "Query response is empty")
```

Comment:
If we use wasmvm errors, maybe for consistency we could use `wasmvmtypes.InvalidResponse` here?

Reply:
super nit: I believe in general the error string that is wrapped always starts with a lowercase letter.

Reply:
done